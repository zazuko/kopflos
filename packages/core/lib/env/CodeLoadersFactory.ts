import type { AnyPointer } from 'clownface'
import LoaderRegistry from 'rdf-loaders-registry'
import cjsLoader from 'rdf-loader-code/ecmaScript.js'
import esmLoader from 'rdf-loader-code/ecmaScriptModule.js'
import { isGraphPointer } from 'is-graph-pointer'

export class CodeLoadersFactory {
  private registry!: LoaderRegistry

  init() {
    this.registry = new LoaderRegistry()
    esmLoader.register(this.registry)
    cjsLoader.register(this.registry)
  }

  load<T>(code: AnyPointer): Promise<T> | T | undefined {
    if (!isGraphPointer(code)) {
      return undefined
    }

    return this.registry.load<T>(code)
  }

  static exports = ['load']
}
