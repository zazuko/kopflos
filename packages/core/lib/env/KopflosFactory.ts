import type { KopflosConfig } from '../Kopflos.js'

export interface KopflosFactory {
  readonly kopflos: {
    readonly config: KopflosConfig
  }
}

export default (config: KopflosConfig) => class implements KopflosFactory {
  get kopflos() {
    return {
      config: Object.freeze(config),
    }
  }

  static exports = ['kopflos']
}
