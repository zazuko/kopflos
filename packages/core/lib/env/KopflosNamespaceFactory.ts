import type { NamespaceBuilder } from '@rdfjs/namespace'
import type { NamespaceFactory } from '@rdfjs/namespace/Factory.js'
import type { Environment } from '@rdfjs/environment/Environment.js'
import type NsBuildersFactory from '@tpluscode/rdf-ns-builders'
import { code } from '@zazuko/vocabulary-extras-builders'

type KopflosTerms = 'api' | 'Config' | 'Api' | 'resourceLoader' | 'handler' | 'method'

declare module '@tpluscode/rdf-ns-builders' {
  interface CustomNamespaces {
    kopflos: NamespaceBuilder<KopflosTerms>
    code: typeof code
  }
}

export class KopflosNamespaceFactory {
  init(this: Environment<NamespaceFactory | NsBuildersFactory>) {
    this.ns = {
      ...this.ns,
      kopflos: this.namespace('https://kopflos.described.at/'),
      code,
    }
  }
}
