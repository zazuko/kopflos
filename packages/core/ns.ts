import rdf from '@zazuko/env-node'

type Properties = 'api' | 'resourceLoader' | 'handler' | 'method' | 'config'
type Classes = 'Config' | 'Api' | 'ResourceShape' | 'Handler'
type Shorthands = 'DescribeLoader' | 'OwnGraphLoader'

export interface KopflosTerms extends Record<Properties | Classes | Shorthands, never>{
}

export const kl = rdf.namespace<keyof KopflosTerms>('https://kopflos.described.at/')
