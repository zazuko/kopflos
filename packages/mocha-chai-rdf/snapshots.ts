/* eslint-disable @typescript-eslint/no-explicit-any */
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import rdf from '@zazuko/env-node'

declare global {
  /* eslint-disable @typescript-eslint/no-namespace */
  namespace Chai {
    interface Assertion {
      canonical: Assertion
    }
  }
}

chai.use(jestSnapshotPlugin())

chai.use((_chai, utils) => {
  const toMatchSnapshot = (chai.Assertion.prototype as any).toMatchSnapshot

  utils.addMethod(chai.Assertion.prototype, 'toMatchSnapshot', function (this: Chai.Assertion, ...args: any[]) {
    const obj = (this as any)._obj

    if (utils.flag(this, 'dataset-canonical')) {
      // Custom behavior for dataset
      expect(rdf.dataset.toCanonical(obj)).toMatchSnapshot(...args)
    } else {
      // Original behavior
      toMatchSnapshot.apply(this, args)
    }
  })

  utils.addProperty(chai.Assertion.prototype, 'canonical', function (this: Chai.Assertion) {
    utils.flag(this, 'dataset-canonical', true)
  })
})
