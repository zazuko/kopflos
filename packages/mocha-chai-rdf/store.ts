/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
