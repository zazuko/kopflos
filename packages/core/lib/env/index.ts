import parent, { Environment } from '@zazuko/env-node'
import { KopflosNamespaceFactory } from './KopflosNamespaceFactory.js'
import { CodeLoadersFactory } from './CodeLoadersFactory.js'

const env = new Environment([
  KopflosNamespaceFactory,
  CodeLoadersFactory,
], { parent })

export default env
export type KopflosEnvironment = typeof env
