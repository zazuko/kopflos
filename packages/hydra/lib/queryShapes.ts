import type { GraphPointer } from 'clownface'
import type { Environment } from '@rdfjs/environment/Environment.js'
import type ClownfaceFactory from 'clownface/Factory.js'
import type NsBuildersFactory from '@tpluscode/rdf-ns-builders'
import { s2q } from '@hydrofoil/shape-to-query'
import { isGraphPointer } from 'is-graph-pointer'
// eslint-disable-next-line import/no-unresolved
import { kl } from '@kopflos-cms/core/ns.js'
import { log } from '@kopflos-cms/core'

interface MemberQueryShapeArgs {
  env: Environment<ClownfaceFactory | NsBuildersFactory>
  collection: GraphPointer
  limit?: number
  offset?: number
}

export function memberQueryShape({ env, collection, limit, offset }: MemberQueryShapeArgs): GraphPointer {
  const { hydra, sh, rdf } = env.ns

  const shape = collection
    .blankNode()
    .addOut(rdf.type, sh.NodeShape)

  const filterShape = collection.blankNode()
  filterShape.addOut(hydra.memberAssertion, collection.out(hydra.memberAssertion))

  function addFilterShape(nodes: GraphPointer) {
    nodes.addOut(sh.filterShape, filterShape)
  }

  function addOrder(orderBy: GraphPointer) {
    return (nodes: GraphPointer) => {
      nodes
        .addOut(sh.orderBy, orderBy)
        .addOut(sh.nodes, addFilterShape)
    }
  }

  function addOffset(offset: number | undefined) {
    const orderBy = collection.out(sh.orderBy)
    let next = addFilterShape
    if (isGraphPointer(orderBy)) {
      next = addOrder(orderBy)
    } else if (limit || offset) {
      log.warn('LIMIT/OFFSET without order by')
    }

    if (offset) {
      return (nodes: GraphPointer) => {
        nodes
          .addOut(sh.offset, offset)
          .addOut(sh.nodes, next)
      }
    }

    return next
  }

  function addLimit(limit: number | undefined) {
    if (limit) {
      return (nodes: GraphPointer) => {
        nodes
          .addOut(sh.limit, limit)
          .addOut(sh.nodes, addOffset(offset))
      }
    }

    return addOffset(offset)
  }

  shape
    .addOut(sh.target, target => {
      target.addOut(rdf.type, s2q.NodeExpressionTarget)
        .addOut(sh.expression, expression => {
          expression.addOut(sh.distinct, addLimit(limit))
        })
    })

  let memberShape = collection.out(kl['hydra#memberShape'])
  if (!isGraphPointer(memberShape)) {
    memberShape = collection.blankNode()
  }
  memberShape
    .addOut(sh.rule, rule => {
      rule
        .addOut(rdf.type, sh.TripleRule)
        .addOut(sh.predicate, hydra.member)
        .addOut(sh.subject, collection.term)
        .addOut(sh.object, sh.this)
    })
  collection.out(hydra.memberAssertion).forEach(memberAssertion => {
    memberShape.addOut(sh.rule, rule => {
      rule
        .addOut(rdf.type, sh.TripleRule)
        .addOut(sh.subject, memberAssertion.out(hydra.subject).term || sh.this)
        .addOut(sh.predicate, memberAssertion.out(hydra.property).term || sh.this)
        .addOut(sh.object, memberAssertion.out(hydra.object).term || sh.this)
    })
  })

  shape.addList(sh.and, memberShape)

  return shape
}

interface TotalsQueryShapeArgs {
  env: Environment<ClownfaceFactory | NsBuildersFactory>
  collection: GraphPointer
}

export function totalsQueryShape({ env, collection }: TotalsQueryShapeArgs) {
  const { hydra, sh, rdf } = env.ns

  const filterShape = collection.blankNode()
  filterShape.addOut(hydra.memberAssertion, collection.out(hydra.memberAssertion))

  return collection
    .blankNode()
    .addOut(rdf.type, sh.NodeShape)
    .addOut(sh.rule, rule => {
      rule
        .addOut(rdf.type, sh.TripleRule)
        .addOut(sh.subject, collection)
        .addOut(sh.predicate, hydra.totalItems)
        .addOut(sh.object, expr => {
          expr
            .addOut(sh.count, count => {
              count.addOut(sh.distinct, distinct => {
                distinct.addOut(sh.filterShape, filterShape)
              })
            })
        })
    })
}
