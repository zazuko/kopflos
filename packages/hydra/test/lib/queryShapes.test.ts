import env from '@zazuko/env-node'
import type { Dataset } from '@zazuko/env/lib/DatasetExt.js'
import { expect, use } from 'chai'
import snapshots from 'mocha-chai-rdf/snapshots.js'
import { createStore } from 'mocha-chai-rdf/store.js'
import type { DatasetCore } from '@rdfjs/types'
import formatsPretty from '@rdfjs-elements/formats-pretty'
// eslint-disable-next-line import/no-unresolved
import { kl } from '@kopflos-cms/core/ns.js'
import { memberQueryShape, totalsQueryShape } from '../../lib/queryShapes.js'
import { ex } from '../../../testing-helpers/ns.js'

env.formats.import(formatsPretty)

describe('@kopflos-cms/hydra/lib/queryShapes.js', () => {
  use(snapshots)

  beforeEach(createStore(import.meta.url, {
    format: 'trig',
    sliceTestPath: [2, -1],
  }))

  describe('memberQueryShape', () => {
    context('unordered collection', () => {
      it('returns shape without limit or offset', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = memberQueryShape({ env, collection })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })
    })

    context('ordered collection', () => {
      it('returns shape with offset', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = memberQueryShape({ env, collection, offset: 5 })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })

      it('returns shape with limit', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = memberQueryShape({ env, collection, limit: 10 })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })

      it('returns shape with limit and offset', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = memberQueryShape({ env, collection, limit: 10, offset: 5 })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })
    })

    context('multiple member assertions', () => {
      it('produces constraint for each', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = memberQueryShape({ env, collection })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })
    })
  })

  describe('totalsQueryShape', () => {
    context('unordered collection', () => {
      it('returns correct total count', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = totalsQueryShape({ env, collection })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })
    })

    context('ordered collection', () => {
      it('returns correct total count', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = totalsQueryShape({ env, collection })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })
    })

    context('multiple member assertions', () => {
      it('returns correct total count', async function () {
        // given
        const collection = this.rdf.graph.namedNode(ex())

        // when
        const shape = totalsQueryShape({ env, collection })

        // then
        expect(await serialize(shape.dataset)).toMatchSnapshot()
      })
    })
  })
})

function serialize(dataset: DatasetCore) {
  return (dataset as Dataset).serialize({
    format: 'text/turtle',
    prefixes: ['sh', 'hydra', 'schema', 'rdf', ['kl', kl().value]],
  })
}
