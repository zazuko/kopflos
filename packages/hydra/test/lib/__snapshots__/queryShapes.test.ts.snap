// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-cms/hydra/lib/queryShapes.js memberQueryShape multiple member assertions produces constraint for each 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g7 hydra:subject <https://ld.admin.ch/district/1004> ;
	hydra:property schema:containsPlace .

_:g6 hydra:property rdf:type ;
	hydra:object schema:PopulatedPlace .

_:g5 hydra:property rdf:type ;
	hydra:object <https://schema.ld.admin.ch/Municipality> .

<http://example.org/> a hydra:Collection ;
	hydra:memberAssertion _:g7, _:g6, _:g5 .

_:b284 a sh:NodeShape ;
	sh:target [
		rdf:type <https://hypermedia.app/shape-to-query#NodeExpressionTarget> ;
		sh:expression [
			sh:distinct [
				sh:filterShape [
					hydra:memberAssertion _:g7, _:g6, _:g5 ;
				] ;
			] ;
		] ;
	] ;
	sh:and (
		[
			sh:rule [
				rdf:type sh:TripleRule ;
				sh:predicate hydra:member ;
				sh:subject <http://example.org/> ;
				sh:object sh:this ;
			], [
				rdf:type sh:TripleRule ;
				sh:predicate schema:containsPlace ;
				sh:subject <https://ld.admin.ch/district/1004> ;
				sh:object sh:this ;
			], [
				rdf:type sh:TripleRule ;
				sh:predicate rdf:type ;
				sh:subject sh:this ;
				sh:object schema:PopulatedPlace ;
			], [
				rdf:type sh:TripleRule ;
				sh:predicate rdf:type ;
				sh:subject sh:this ;
				sh:object <https://schema.ld.admin.ch/Municipality> ;
			] ;
		]
	) .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js memberQueryShape ordered collection returns shape with limit 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g4 sh:path schema:name .

_:g3 hydra:object <https://schema.ld.admin.ch/Municipality> ;
	hydra:property rdf:type .

<http://example.org/> a hydra:Collection ;
	sh:orderBy _:g4 ;
	hydra:memberAssertion _:g3 .

_:b261 a sh:NodeShape ;
	sh:target [
		rdf:type <https://hypermedia.app/shape-to-query#NodeExpressionTarget> ;
		sh:expression [
			sh:distinct [
				sh:limit 10 ;
				sh:nodes [
					sh:orderBy _:g4 ;
					sh:nodes [
						sh:filterShape [
							hydra:memberAssertion _:g3 ;
						] ;
					] ;
				] ;
			] ;
		] ;
	] ;
	sh:and (
		[
			sh:rule [
				rdf:type sh:TripleRule ;
				sh:predicate hydra:member ;
				sh:subject <http://example.org/> ;
				sh:object sh:this ;
			], [
				rdf:type sh:TripleRule ;
				sh:predicate rdf:type ;
				sh:subject sh:this ;
				sh:object <https://schema.ld.admin.ch/Municipality> ;
			] ;
		]
	) .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js memberQueryShape ordered collection returns shape with limit and offset 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g4 sh:path schema:name .

_:g3 hydra:object <https://schema.ld.admin.ch/Municipality> ;
	hydra:property rdf:type .

<http://example.org/> a hydra:Collection ;
	sh:orderBy _:g4 ;
	hydra:memberAssertion _:g3 .

_:b272 a sh:NodeShape ;
	sh:target [
		rdf:type <https://hypermedia.app/shape-to-query#NodeExpressionTarget> ;
		sh:expression [
			sh:distinct [
				sh:limit 10 ;
				sh:nodes [
					sh:nodes [
						sh:orderBy _:g4 ;
						sh:nodes [
							sh:filterShape [
								hydra:memberAssertion _:g3 ;
							] ;
						] ;
					] ;
					sh:offset 5 ;
				] ;
			] ;
		] ;
	] ;
	sh:and (
		[
			sh:rule [
				rdf:type sh:TripleRule ;
				sh:predicate hydra:member ;
				sh:subject <http://example.org/> ;
				sh:object sh:this ;
			], [
				rdf:type sh:TripleRule ;
				sh:predicate rdf:type ;
				sh:subject sh:this ;
				sh:object <https://schema.ld.admin.ch/Municipality> ;
			] ;
		]
	) .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js memberQueryShape ordered collection returns shape with offset 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g4 sh:path schema:name .

_:g3 hydra:object <https://schema.ld.admin.ch/Municipality> ;
	hydra:property rdf:type .

<http://example.org/> a hydra:Collection ;
	sh:orderBy _:g4 ;
	hydra:memberAssertion _:g3 .

_:b250 a sh:NodeShape ;
	sh:target [
		rdf:type <https://hypermedia.app/shape-to-query#NodeExpressionTarget> ;
		sh:expression [
			sh:distinct [
				sh:offset 5 ;
				sh:nodes [
					sh:orderBy _:g4 ;
					sh:nodes [
						sh:filterShape [
							hydra:memberAssertion _:g3 ;
						] ;
					] ;
				] ;
			] ;
		] ;
	] ;
	sh:and (
		[
			sh:rule [
				rdf:type sh:TripleRule ;
				sh:predicate hydra:member ;
				sh:subject <http://example.org/> ;
				sh:object sh:this ;
			], [
				rdf:type sh:TripleRule ;
				sh:predicate rdf:type ;
				sh:subject sh:this ;
				sh:object <https://schema.ld.admin.ch/Municipality> ;
			] ;
		]
	) .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js memberQueryShape unordered collection returns shape without limit or offset 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g1 sh:property [
		sh:path schema:name ;
	] ;
	sh:rule [
		rdf:type sh:TripleRule ;
		sh:predicate hydra:member ;
		sh:subject <http://example.org/> ;
		sh:object sh:this ;
	], [
		rdf:type sh:TripleRule ;
		sh:predicate rdf:type ;
		sh:subject sh:this ;
		sh:object <https://schema.ld.admin.ch/Municipality> ;
	] .

_:g0 hydra:object <https://schema.ld.admin.ch/Municipality> ;
	hydra:property rdf:type .

<http://example.org/> a hydra:Collection ;
	kl:memberShape _:g1 ;
	hydra:memberAssertion _:g0 .

_:b242 a sh:NodeShape ;
	sh:target [
		rdf:type <https://hypermedia.app/shape-to-query#NodeExpressionTarget> ;
		sh:expression [
			sh:distinct [
				sh:filterShape [
					hydra:memberAssertion _:g0 ;
				] ;
			] ;
		] ;
	] ;
	sh:and (
		_:g1
	) .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js totalsQueryShape multiple member assertions returns correct total count 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g7 hydra:subject <https://ld.admin.ch/district/1004> ;
	hydra:property schema:containsPlace .

_:g6 hydra:property rdf:type ;
	hydra:object schema:PopulatedPlace .

_:g5 hydra:property rdf:type ;
	hydra:object <https://schema.ld.admin.ch/Municipality> .

<http://example.org/> a hydra:Collection ;
	hydra:memberAssertion _:g7, _:g6, _:g5 .

_:b351 a sh:NodeShape ;
	sh:rule [
		rdf:type sh:TripleRule ;
		sh:subject sh:this ;
		sh:predicate hydra:totalItems ;
		sh:object [
			sh:count [
				sh:distinct [
					sh:filterShape [
						hydra:memberAssertion _:g7, _:g6, _:g5 ;
					] ;
				] ;
			] ;
		] ;
	] .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js totalsQueryShape ordered collection returns correct total count 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g3 hydra:object <https://schema.ld.admin.ch/Municipality> ;
	hydra:property rdf:type .

<http://example.org/> a hydra:Collection ;
	sh:orderBy [
		sh:path schema:name ;
	] ;
	hydra:memberAssertion _:g3 .

_:b345 a sh:NodeShape ;
	sh:rule [
		rdf:type sh:TripleRule ;
		sh:subject sh:this ;
		sh:predicate hydra:totalItems ;
		sh:object [
			sh:count [
				sh:distinct [
					sh:filterShape [
						hydra:memberAssertion _:g3 ;
					] ;
				] ;
			] ;
		] ;
	] .

"
`;

exports[`@kopflos-cms/hydra/lib/queryShapes.js totalsQueryShape unordered collection returns correct total count 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix hydra: <http://www.w3.org/ns/hydra/core#> .
@prefix schema: <http://schema.org/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix kl: <https://kopflos.described.at/> .

_:g0 hydra:object <https://schema.ld.admin.ch/Municipality> ;
	hydra:property rdf:type .

<http://example.org/> a hydra:Collection ;
	kl:memberShape [
		sh:property [
			sh:path schema:name ;
		] ;
	] ;
	hydra:memberAssertion _:g0 .

_:b296 a sh:NodeShape ;
	sh:rule [
		rdf:type sh:TripleRule ;
		sh:subject sh:this ;
		sh:predicate hydra:totalItems ;
		sh:object [
			sh:count [
				sh:distinct [
					sh:filterShape [
						hydra:memberAssertion _:g0 ;
					] ;
				] ;
			] ;
		] ;
	] .

"
`;
